[DOCS-9] Time Methods Draft Doc--[DO NOT MERGE]

rajkaramchedu opened the pull request:
This is the start of the new draft document for time-related methods. I am just getting started, and this is a work in progress (WIP), and the flow and structure may change before the final version.

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+.. _smartapp-time-methods:
+
+Time Methods
```
Maybe make the title of this document more task-oriented, something like "Working with Time"?

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+For example, a SmartApp can turn on a room light when a door is opened but only during certain hours, or wake up the house in the morning at different times based on what day of the week it is.
+
+The SmartThings platform provides a few time methods that can readily be used in a SmartApp to accomplish such automations.
+These time methods support a variety of time-related queries such as, to name a few, get the current time, find out today's date, know the time zone, or find out if a given moment of time is between a preset time-window.
```
Minor, but maybe tighten this sentence up a bit. Perhaps we can remove certain phrases like "to name a few" as a start.

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+The SmartThings platform provides a few time methods that can readily be used in a SmartApp to accomplish such automations.
+These time methods support a variety of time-related queries such as, to name a few, get the current time, find out today's date, know the time zone, or find out if a given moment of time is between a preset time-window.
+
+timeOfDayIsBetween()
```
Let's make this section title more task-oriented instead of the method name. Something like "Taking Action Within a Time Window" (or something to that effect). That allows the reader to understand the use case, and then understand the APIs involved in solving that particular task, rather than starting at the API level and working backwards.

(this isn't something we do a great job of consistently in the docs today, but is something we've learned and are trying to get better at :))

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+These time methods support a variety of time-related queries such as, to name a few, get the current time, find out today's date, know the time zone, or find out if a given moment of time is between a preset time-window.
+
+timeOfDayIsBetween()
+^^^^^^^^^^^^^^^^^^^^
```
Top-level sections should have `----` instead of `^^^^^` (which would be for subsections).

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+timeOfDayIsBetween()
+^^^^^^^^^^^^^^^^^^^^
+
+One such time-related method is ``timeOfDayIsBetween()``.
```
I think we can remove this sentence and just dive into the use case.

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+Let's see how this ``timeOfDayIsBetween()`` method comes in handy to set up a SmartApp that accomplishes such an automation.
+
+Refer to the SmartApp code below.
+First we set up the ``preferences()`` section with ``openCloseSensor``, a open/close sensor that detects when the door is opened, and a ``roomLight`` that controls the switch to the room light.
```
"an open/close sensor"

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+
+    }
+
+
```
Need to clean up the formatting/indention here. That's usually caused by pasting code from the IDE, which unfortunately mangles indentation with a mix of tabs and spaces :(

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+            // Door is opened. Now check if the current time is within the visiting hours window
+
+               def between = true
+               between = timeOfDayIsBetween(fromTime, toTime, new Date(), location.timeZone)
```
Don't we need to handle the case where a location has a null time zone? Unless you've discovered otherwise, we should probably show an example similar to our own SmartApps that will use a time zone from either a user preference or some other default, if the location timeZone is null.

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+               between = timeOfDayIsBetween(fromTime, toTime, new Date(), location.timeZone)
+
+              if (between) {
+             		def currTime = now()
```
What is the purpose of the `currTime` variable here?

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+                roomLight.off()
+
+              }
+    }
```
Same formatting/indentation issue with this code block.

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+              }
+    }
+
+The ``timeOfDayIsBetween()`` method returns a Boolean ``true`` if the current time and date, i.e., ``new Date()``, is within the time-window specified by ``fromTime`` and ``toTime`` arguments, and returns ``false`` if the ``new Date()`` value is outside the bounds of ``fromTime`` and ``toTime``.
```
If it always returns `true` or `false`, why do we set up a variable (`between`) in the code example and initialize it to true? Couldn't we do `def between = timeOfDayIsBetween(...)`? (Maybe there's a reason or case that we can't, I don't know, just as a reader would wonder that.)

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+
+
+.. note::
+  Note that the ``timeOfDayIsBetween()`` resolves to the minute. This means that, for example, if the ``toTime`` argument to it is, say, ``16:14`` (4:14pm), then the time-window is closed before 4:14:59, and *not* before 4:13:59.
```
This seems like pretty important information, and a potential gotchya. I had to think about this a bit and read the sentence a couple times to get it though. "resolves to the minute" doesn't immediately make sense to me, because I would think it all is minute-based since user inputs cannot be more granular than a minute. But wording aside, if we can use less words to describe this, and then supplement with a few clear examples, that might be much clearer. I could see even listing a few cases in tabular form, with columns for "start time", "end time", "result". Then we could quickly/clearly see how it works.

I think that information should be in the API reference documentation for `timeOfDayIsBetween()` (either as well as, or just instead of here), since it is pretty important for users to understand that. 

We should also link to the API reference docs from the guide, a good rule-of-thumb is that the first reference to the method is a link to the reference docs, and the remaining usages can use the code literal formatting.

jimmyjames commented on smartapp-developers-guide/time-methods.rst:
```diff
@@ -0,0 +1,83 @@
+^^^^^^^^^^^^^^^^^^^^
+
+One such time-related method is ``timeOfDayIsBetween()``.
+A meaningful home automation with SmartThings is to turn on a room light when the door is opened between certain hours, and do not turn on the light during other times.
```
"meaningful" is a bit subjective; maybe just say "A common automation is to turn on a light when a door is opened, but only between certain hours." Or something a bit more concise.

jimmyjames commented:
Also this guide will need to be added to the smartapp-developers-guide/index.rst so it is included in the ToC.

rajkaramchedu commented:
Actually, can we delete this PR? I suspect there are a few "not the latest" issues with this branch on my side, and besides we've addressed all these comments in PRs #457 and #456 and I'd like to not keep this branch.

jimmyjames commented:
Sure, go ahead and close this PR using the "Close and comment" button here.

rajkaramchedu commented:
Closing this PR. Refer to PRs #457 and #456.
